File: leetcode-submissions/63_Unique_Paths_II.cpp
class Solution {
public:
vector<vector<int>>grid;

int f(int i,int j){
    //now setting base cases
    int n=grid.size();
    int m=grid[0].size();
    if(i>=n or j>=m or i<0 or j<0){
        return 0;
    }
    if(i==n-1 and j==m-1){
        return 1;
    }
    if(grid[i][j]==1){
        return 0;
    }
    return f(i,j+1)+f(i+1,j);

    
}
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
        grid=obstacleGrid;
        return f(0,0);
    }
};

File: leetcode-submissions/63_Unique_Paths_II.ts
function uniquePathsWithObstacles(grid: number[][]): number {
    const m :number = grid.length , n = grid[0].length;
    const dp : Array<number[]> = Array.from({length:m+1},()=>Array.from({length:n+1}, ()=>-1));
    return solve(m , n , 0 , 0 , dp, grid);
};
const solve=(m :number , n :number , i :number , j:number ,  dp:Array<number[]>, grid:Array<number[]>) =>{
    if(i==m-1  && j==n-1){
        return grid[i][j]==0;

    }
    
    if(i<0 || j<0 || j>=n || i>=m  || grid[i][j]==1){
        return 0 ;
    }
    if(dp[i][j]!=-1){
        return dp[i][j];
    }
    return dp[i][j]=  solve(m , n , i, j+1, dp, grid)+solve(m , n , i+1, j, dp, grid);
}